Sources/Network/NetworkAuthenticator.swift

```diff
@@ -0,0 +1,29 @@
+//
+//  Authenticator.swift
+//  Alicerce
+//
+//  Created by Luís Portela on 03/07/2017.
+//  Copyright © 2017 Mindera. All rights reserved.
+//
+
+import Foundation
+
+public protocol NetworkAuthenticator {
+    func authenticate(request: URLRequest, _ completion: @escaping (URLRequest) -> Void)
+
+    func shouldRetry(with data: Data?, response: HTTPURLResponse?, error: Swift.Error?) -> Bool
```

Comment:
No.

Why use and abuse of `associatedType` and add an unnecessary complexity. This is called `NetworkAuthenticator`, if this was called `EverythingAuthenticator`, i would agree with that.
But please explain me the advantage of add that complexity.